L1Trigger/TrackerDTC/plugins/TrackerDTCProducer.cc

```diff
@@ -0,0 +1,176 @@
+  };
+
+  Producer::Producer(const ParameterSet& iConfig) : settings_(iConfig) {
+    // book in- and output ED products
+    getTokenTTStubDetSetVec_ = consumes<TTStubDetSetVec>(settings_.inputTagTTStubDetSetVec());
+    putTokenTTDTC_ = produces<TTDTC>(settings_.productBranch());
+    // book ES products
+    getTokenTTStubAlgorithm_ =
+        esConsumes<TTStubAlgorithm<Ref_Phase2TrackerDigi_>, TTStubAlgorithmRecord, Transition::BeginRun>(
+            settings_.inputTagTTStubAlgorithm());
+    getTokenMagneticField_ =
+        esConsumes<MagneticField, IdealMagneticFieldRecord, Transition::BeginRun>(settings_.inputTagMagneticField());
+    getTokenTrackerGeometry_ = esConsumes<TrackerGeometry, TrackerDigiGeometryRecord, Transition::BeginRun>(
+        settings_.inputTagTrackerGeometry());
+    getTokenTrackerTopology_ =
+        esConsumes<TrackerTopology, TrackerTopologyRcd, Transition::BeginRun>(settings_.inputTagTrackerTopology());
+    getTokenCablingMap_ =
+        esConsumes<TrackerDetToDTCELinkCablingMap, TrackerDetToDTCELinkCablingMapRcd, Transition::BeginRun>(
+            settings_.inputTagCablingMap());
+  }
+
+  void Producer::beginRun(const Run& iRun, const EventSetup& iSetup) {
+    // read in detector parameter
+    settings_.setTrackerGeometry(&iSetup.getData(getTokenTrackerGeometry_));
+    settings_.setTrackerTopology(&iSetup.getData(getTokenTrackerTopology_));
+    settings_.setMagneticField(&iSetup.getData(getTokenMagneticField_));
+    settings_.setCablingMap(&iSetup.getData(getTokenCablingMap_));
+    settings_.setTTStubAlgorithm(iSetup.getHandle(getTokenTTStubAlgorithm_));
+    settings_.setProcessHistory(iRun.processHistory());
+    // convert data fromat specific stuff
+    settings_.beginRun();
+    // check coniguration
+    settings_.checkConfiguration();
+    // convert cabling map
+    settings_.convertCablingMap();
+    // convert outer tracker geometry
+    const vector<DetId>& cablingMap = settings_.cablingMap();
+    auto acc = [](int& sum, const DetId& detId) { return sum += !detId.null(); };
+    const int numModules = accumulate(cablingMap.begin(), cablingMap.end(), 0, acc);
+    modules_.reserve(numModules);
+    int modId(0);
+    for (const DetId& detId : cablingMap) {
+      if (!detId.null())
+        modules_.emplace_back(&settings_, detId, modId);
+      modId++;
+    }
+    settings_.setModules(modules_);
+  }
+
+  void Producer::produce(Event& iEvent, const EventSetup& iSetup) {
+    // empty DTC product
+    TTDTC product(settings_.numRegions(), settings_.numOverlappingRegions(), settings_.numDTCsPerRegion());
+
+    // read in stub collection
+    Handle<TTStubDetSetVec> handleTTStubDetSetVec;
+    iEvent.getByToken(getTokenTTStubDetSetVec_, handleTTStubDetSetVec);
+
+    // apply cabling map
+    auto acc = [](int& sum, const TTStubDetSet* module) { return sum += module ? module->size() : 0; };
+    static const vector<const TTStubDetSet*> nullDTC(settings_.numModulesPerDTC(), nullptr);
```

Comment:
@makortel is this usage fully thread safe?

Reply:
In principle it is if `settings_.numModulesPerDTC()` is a constant and does not change from one call to the next.

On the other hand, the `nullDTC` seems to be used only to initialized `ttDTCs` on the line below, and I can't really think of any practical benefit from `nullDTC` being `static const`. The line below could be just
```cpp
vector<vector<const TTStubDetSet*>> ttDTCs(settings_.numDTCs(), vector<const TTStubDetSet*>(settings_.numModulesPerDTC(), nullptr));
```